### ISSUE TITLE
Closures are bad?

### ISSUE THREAD

**Issue by keean** [keean/zenscript#33]
I have been writing some application code in JavaScript and TypeScript recently, and one of the things that I have started to realise is that using closures results in lots of garbage for the collector to clean up. It is becoming apparent to me that the convenience of closures costs you in performance. The easy solution is to refactor the captured variables as object properties, and the closure as an object property. A lot of the time we do not need the 'queued state' that is provided by closures, and when we do, creating a new object to encapsulate that state is not a big problem, especially when you have anonymous objects. This reminds me that 'runnable' objects in Java or 'function objects' in C++ serve the same purpose as closures.

I guess my thoughts are should the syntax make closures harder to use (by forcing you to create a function object) which then makes the programming effort align with the execution efficiency. To put this another way closures make it easy for developers to write inefficient code, by removing them from the language syntax developers would fall back to creating object properties and methods for the callbacks which is more efficient, and would only create a new object to hold the values where it was really necessary. This is approach could be described as "make the easy way the correct/efficient way". I don't know if this is something I really want to do yet, its here more as an observation, and an idea about how syntax can support the programmer doing the right thing.

**Comment by shelby3** [comment]
From my perspective, you’re first paragraph [lacks clarity](https://github.com/keean/zenscript/issues/30#issuecomment-302321411). I am not inside your head. I need you to explain in sufficient detail so as to convey holistically what you are thinking.

From my perspective, a closure includes a reference to its bound variables (aka environment) and thus the bound variables can only be garbage collected or freed with reference counting. I do not understand what you propose otherwise? You provided insufficient explanation (examples) of how “object properties” would improve this.

I am thinking that only by keeping allocations ordered on the stack, can we free them deterministically. The callbacks in a `Promise` can actually be implemented with green threads so the stack ordering can be respected and thus no closures are necessary.

**Comment by keean** [comment]
The problem is the creation of garbage, consider:
```
doit(() => {
    somevar =
}
```
Every time doit is called a new closure is created storing the captured variables on the heap. If this happens in a game or animation, or on any frequent event like a mouse-event, this creates new garbage which eventually triggers a collection that pauses the interactivity. So the less garbage we create the less "jank" there will be in the UI. Most times the above can be replaced with a new method (TypeScript):
```
private readonly somevar = 0;
public my_callback() {
    this.somevar = ...
}
```
and called like this:
```
doit(my_callback);
```
This does not heap allocate on the callback, and so generates less garbage.

To be clear we are talking about the rate of garbage generation, and the pauses caused by the GC dealing with the garbage.

**Comment by shelby3** [comment]
JavaScript should be reusing the (and not be creating new) instances of the variables in the closure environment on each closure when calling `doit`. Because you are creating a new instance of the callback function each time you call `doit`, thus a new closure environment is created each time.

Thus afaics, you do not need an object to solve the problem. You need to create one instance of the closure and reuse it when you call `doit`, which is what you end up achieving with the afaics unnecessary burden of the introducing the class object.

And note that making the variables object properties did not guarantee they are unboxed in JavaScript.

**Comment by keean** [comment]
My point above is that closures make it too easy to do this.

**Comment by shelby3** [comment]
Does it have anything to do with closures? Even if the variables are local, a new instance of the callback function is going to be created on the heap each time `doit` is called if the programmer is not careful.

**Comment by keean** [comment]
It's creating the closures on the heap that is creating the garbage.

**Comment by shelby3** [comment]
I have stated that even without closures:

> a new instance of the callback function is going to be created on the heap each time doit is called if the programmer is not careful.

Have you refuted that claim?

**Comment by shelby3** [comment]
@shelby3 wrote:
> JavaScript should be reusing the (and not be creating new) instances of (at least copy-by-reference) variables in the closure environment on each closure when calling `doit`. Because you are creating a new instance of the _inline_ callback function each time you call your first `doit` example, thus a new closure environment is created each time (but not new instances of the at least copy-by-reference variables in the closure).

It is probably true that in general cases when the referenced variable would otherwise be stored on the stack (i.e. copy-by-value), then referencing it in a closure will force it to be placed on the heap instead. However, I suspect though that JavaScript only uses copy-by-value semantics on some primitive types such as `Number`.

> And note that making the variables object properties did not guarantee they are unboxed in JavaScript.

Object properties may or may not solve the above problem, depending on if JavaScript always unboxes those primitive types that are copy-by-value on the stack.

> Thus afaics, you do not need object properties to solve the problem. You need to create one instance of the closure (and callback) and reuse it when you call `doit`, which is what you end up achieving with the afaics unnecessary burden of the introducing the class object.

But in any case, it is the multiple instances of the function (especially inside a loop) that amplify the problem the most.

I apologize if that all was not clear from what I wrote before as quoted above.

**Comment by keean** [comment]
I am reporting a real problem in a real application, so there is no question it is a real problem. I have had to refactor the code to get acceptable performance.
Perhaps I have not sufficiently explained, but creating the closure creates an object on the heap, referring to member variables and methods of the object hosting the callback does not (because we create no new objects on the heap). Note: I am not creating a new object for each callback, it's not necessary because in JavaScript only one event handler can be called at a time because it is single threaded.

**Comment by shelby3** [comment]
> but creating the closure creates an object on the heap

Instancing a function also creates an object on the heap. Unless I see your code, I can not trust that you know what is causing your problem and trust that you have analyzed it correctly.

> referring to member variables and methods of the object hosting the callback does not

I am sorry but it seems like you have entirely not understood what I wrote.

The object hosting the callback is on the heap. So you instanced the callback only once (and it is on the heap) and that explains the reduction in GC load.

>  Note: I am not creating a new object for each callback, it's not necessary because in JavaScript only one event handler can be called at a time because it is single threaded.

This is the first time you have mentioned event handler. You leave out so many details then expect me to trust some nebulous bug report. That is the not correct way to report a bug. You must provide a reproducible code sample that can be independently analyzed.

How in the heck could the GC load be impacted at all if you are not instancing the callback for the event handler many times?

**Comment by shelby3** [comment]
> My thought is to not allow variable capture by a nested function

That is analogous to killing the dog to kill the fleas.

There are valid use cases of inner functions with closures, e.g. partial application.

**Comment by shelby3** [comment]
> If you read this paper:
> 
> https://cseweb.ucsd.edu/~goguen/pps/utyop.ps
> 
> It makes a good argument for not allowing first class functions at all, and instead provides modules with sorts and subsorts. Apparently there is some conflict trying to provide higher order functions and sub-sorts in the same system, and the author of this paper believes that sub-sorts are more useful for programming than higher order functions. I am not saying I agree, but I need to think about it.

I found this:

https://futhark-lang.org/blog/2017-01-25-futhark-module-system.html#parametric-modules

Module composition is all done at compile-time, thus it is a total order and can not be dynamic open modularity.

**Comment by keean** [comment]
> How in the heck could the GC load be impacted at all if you are not instancing the callback for the event handler many times? Meaning that it was instancing the callback only once that was likely responsible for the reduction in GC load, and very little if nothing to do with closures.

Back near the beginning I said:

> Every time doit is called a new closure is created storing the captured variables on the heap. If this happens in a game or animation, or on any frequent event like a mouse-event, this creates new garbage which eventually triggers a collection that pauses the interactivity. 

This says that the function is called from an animation loop or event handler.

```
var thing
list.forEach(() => thing++)
for(var o in list) doit(() => thing++)
```
It would be much more efficient to write:
```
var thing,
    f = () => thing++
list.forEach(() => thing++) // this one only creates one instance of the inline function
for(var o in list) doit(f)
```
Yes you seem to have understood the problem, except instead of a loop we are talking about a callback or an event handler. Consider:
```
onmousemove = (event) => {
    setTimeout(() => { ... }, 1000)
}
```
now consider this is called as part of the 'mousemove' event handler, it will create hundreds of new closures per second. Instead we want to write:
```
tfn() {
   ...
}

onmousemove(event) {
   setTimeout(this.tfn, 1000)
}
```
This form creates much less garbage.

**Comment by shelby3** [comment]
And precisely because you are not creating a new instance of the function each time. It has nothing to do with the closure.

**Comment by keean** [comment]
> Yes I entirely understood the problem. But for some unknown reason, you apparently do not read (carefully) what I write.

I don't understand why you are disagreeing with me, the first version clearly creates a closure on the heao _every_time_ the event handler is called. It is the accumulation of these closures on the heap that is causing the problem. I know this is the case, because when I stop the closure being created every time the event handler is called the problem goes away. 

I think your point is that you can have a closure elsewhere (outside of the loop as you put it), so it is not the closure that is the problem, but the placement of the closure. But that misses my point that having closures makes it easy to code in this way. Remove closures from the language and now the programmer is forced to do it a different way which does not create so much garbage.

So the whole point of my comment was to consider the merit of choosing language features that enable good coding only, rather than having features that enable both good and bad coding styles.

**Comment by shelby3** [comment]
My point is that it is creating a new instance of the function on each event callback in your example which is causing a object (storing the function and optionally its closure environment) to be allocated on the heap, thus placing more workload on the GC. I had explained in great detail why the addition of a closure environment is not the significant additional heap allocation overhead. The significance of the heap allocation is caused by the creation of multiple instances of the inline function (as the callback) in your example. I hope that is clear by now, as I do not know how else to write that in plain English. I feel this is about the 3rd or 4th time I have written that in this thread. Do I have some error in my conceptualization which is cause you’re (what appears to me to be) cognitive dissonance?

<sup>I do not wish to be derogatory in any way. I appreciate your willingness to interact with me. It is a bit frustrating that we talk past each other so often, but I guess that it is good that our brains do not think exactly the same, so that we can catch each others’ errors and blind spots. I am under extreme time pressure right now, so it is impacting my patience for repeating myself, but as @skaller wrote in the Typeclass objects thread, humans need redundancy of communication.</sup>

---------------

Btw, I thought about the nature of this bug and I realized a generalized improvement which can help prevent the bad programming without removing the generality we need for good programming.

The bad cases where we create many instances of function needlessly can only occur for instances that would not otherwise be automatically deallocated when the stack frame is if we have an [RAII model you proposed and I agreed with](https://github.com/keean/zenscript/issues/32#issuecomment-301041309) (see [also](https://github.com/keean/zenscript/issues/17#issuecomment-301312332)).

In other words, if our language distinguished instances that could be deterministically deallocated when the stack unwinds, it could identify those allocations which are candidates for unnecessarily overloading the GC. In fact, the candidates for creating the problem are not only function constructions, but also include any object creations that can not be deallocated when the stack frame is.

The language could either force these allocations to have a special annotation and/or the IDE could syntax color them conspicuously, so the programmer is aware these must be studied more closely.

**Comment by keean** [comment]
> The significance of the heap allocation is caused by the creation of multiple instances of the inline function (as the callback) in your example. 

You are right it is the rate of object creation that is the problem. However here a 'functional' idiom is responsible for the high rate of object creation (closures are implicit objects).

The inline function code takes no space, the code is only there once in the source, and in the machine code. A closure is a pointer to the function code and a pointer to the environment containing the state. A closure object is created whether the environment is empty or not, so the environment pointer can be null. For every call to the inline function a new closure object is created pointing to the same machine code for the function. So the garbage created consists of closure objects.

**Comment by shelby3** [comment]
> The inline function code takes no space, the code is only there once in the source, and in the machine code. A closure is a pointer to the function code and a pointer to the environment containing the state.

Actually you are correct that if there is no closure environment, then there is no reason to create a new instance of the function code on each call of your event handler.

And even if there is a closure environment, if all the variables in that environment are pass-by-reference (as opposed to pass-by-value) then all the instances (all of the calls to your event handler) can share the same closure environment.

So in theory it should only be the closure on pass-by-value variables that would cause the excessive heap allocations. And you are correct that is an implicit and insidious problem that the programmer would not be aware of.

**Comment by shelby3** [comment]
Okay so JavaScript’s lack of static checks and runtime dynamism with `eval` creates a nasty corner case where a defining (and referencing) a function within another function creates a closure environment on the heap for the former every time the latter said function is called **even if the former said function does not statically reference any variables in the lexical environment** external to itself. <sup>That is a rather insidious corner case that probably explains why my computer runs out-of-memory when I have long running JavaScript GUI applications running in my browser.</sup>

If we have static analysis with our proposed language compiler, then we can fix that problem automatically by creating such functions in an external (global) scope automatically, so the programmer does not have to worry about unwittingly creating this bug.

But it also means in our language any concept like `eval` needs to be recompiled with variable source code to be run, which we need anyway because our language is statically compile. Remember I also wrote in a different Issues thread that the caller of dynamically compiled code should be able to restrict its access by limiting the imports in scope for the compilation, in order to try to prevent script injection attacks.

Yet the problem remains for closures (i.e. that are not empty as statically compiled) they are allocated on the heap and the GC is tasked which freeing them which is a burden. In my mind, GC should be more for long-lived objects and not for function local variables, which is why I like the RAII concept and deterministically deallocated function (and/or block) local variables.

My [recently proposed](https://github.com/keean/zenscript/issues/32#issuecomment-301060409) “stored (transitive)” annotation will help alleviate this issue. When a reference to a callback is not annotated as stored, then the caller knows no references are held on the callback so the closure can be deallocated deterministically. We could even eventually use a stack (even within JavaScript as Emscripten does employing `ArrayBuffer`).

Perhaps closure cases that are not resolved by the above improvements need to be conspicuously annotated (to remove a compiler warning), so the programmer is aware of the danger of the inefficiency and potential to thrash the GC.

**Comment by shelby3** [comment]
@shelby3 wrote:
> @keean wrote:
> > … the first version clearly creates a closure on the heap _every_ time the event handler is called …
> 
> Referring presumably to this example:
> 
> > ```
> > onmousemove = (event) => {
> >     setTimeout(() => { ... }, 1000)
> > }
> > ```

…

> Ah maybe each nested lexical scope is an `Object` with the parent lexical scopes in the prototype chain! So that would solve the shadowing.
> 
> Edit#4: even though all function local variables have to be on the heap, they could still be unboxed in the closure environment object.

I am contemplating that it is key to note that passing as an argument (to `setTimeout`) a function declared within the local lexical scope, passes a reference for the local lexical scope `Object` to the said function. Since `setTimeout` stores that reference, then the local lexical scope `Object` (containing all the arguments and local variables of the parent function) can not be freed by any generational GC scheme. <sub>(Also tangentially as I had noted before that given the absence of read-only annotations on arguments, JavaScript must assume the said function stores the said reference and thus can not do any static optimization.)</sub>

This is important because I presume that optimized JavaScript engines such as V8 are normally proficient at rapid generation collection of those lexical scope `Objects` which are essentially forming a linked-list form of a stack.

So really the problem you found is related to the nature of `setTimeout` taking a long-duration reference to the input argument than it is specific to closures.

It is very important to know if my presumptions in this post are correct, because impacts the design of the language we create if we are targeting JavaScript. Specifically, I am contemplating a `foreach` typeclass method on iterators and I am trying to decide if there is any advantage to it taking an extra input which it threads to its callback so as to provide a means for avoiding passing such input via closures, so that the callback could be moved to an external lexical scope.

Thus I am presuming that closures are not as bad in JavaScript as you were thinking and JavaScript is not as derelict and inefficient as I was thinking. Instead it seems the only issue is storing a reference for longer than the current function (thus bypassing the generational GC scheme) of a object created in frequently called function or loop.

Feedback?

**Comment by keean** [comment]
Yes, JavaScript does okay when objects get created and destroyed quickly.  However the problem is not specific to get timeout and simple assignments of closures to member variables can cause the same effect. It is always better to create a named function like this:
```
var myfn = (x) => x + 1

... in loop
this.callback = myfn
...
```
Than use an anonymous one:
```
in loop...
this.callback = (x) => x + 1
```
But it's not as bad as with setTimeout, which has the added problem of creating lots of timer objects as well as function objects.

**Comment by shelby3** [comment]
But again that problem is not specific to closure, but rather that creating any object in a loop or frequently and storing a reference to it in a longer-lived instance will thus bypass the generational GC.

I stumbled onto this resource and do not feel well enough to read it all and assimilate it, so I do not know if it is a good read:

https://github.com/getify/You-Dont-Know-JS/tree/master/scope%20%26%20closures

**Comment by shelby3** [comment]
I [wrote](https://github.com/keean/zenscript/issues/33#issuecomment-302717983):
> Even though a generational GC helps, it still is (afaik) inferior in efficiency to deterministic stack frame deallocation.

Ah this may not be true. A generational GC may be able to group (and thus amortize some costs for) deallocation into larger economies-of-scale.

**Comment by shelby3** [comment]
[Mitigating the potential memory leaks with closures](https://github.com/keean/zenscript/issues/35#issuecomment-355297857).

**Comment by shelby3** [comment]
### Closure Typing For Imperative Languages

Closures over (even immutable [except in some cases](https://realworldocaml.org/v1/en/html/imperative-programming-1.html#relaxing-the-value-restriction)) state cause a complication to the type system to avoid [breaking type safety](http://mlton.org/ValueRestriction#_why_the_value_restriction_exists).

ML [settled](http://mlton.org/ValueRestriction#_alternatives_to_the_value_restriction) on the [_value restriction_](http://mlton.org/ValueRestriction) which restricts parametric polymorphism even in cases which would be type safe.

**Taking the perspective that the typing problem [originates from closure over state](https://cstheory.stackexchange.com/questions/38173/why-does-mls-value-restriction-stop-you-from-capturing-parametric-values)** (although not all closures-over-state cause the problem as noted below for covariance and [OCaml’s relaxation for immutability](https://realworldocaml.org/v1/en/html/imperative-programming-1.html#relaxing-the-value-restriction)), A significantly less restrictive [_Closure Typing_](http://mlton.org/References#Wright95) (which restricts less than the less generalized [_Weak Types_](http://mlton.org/References#Wright95)) restricts the type parameters for any state in a closure to the instantiated concrete types at the point of evaluation that creates the closure. IOW, [for example the partial application](https://realworldocaml.org/v1/en/html/imperative-programming-1.html#partial-application-and-the-value-restriction) of any input of a curried function<sup>1</sup> creates a closure over parametrised state causing that type parameter to be concrete from that point for in the evaluation. _Closure Typing_ [was dropped when](https://stackoverflow.com/questions/22507448/the-value-restriction#comment34244466_22507665) the ML functors were added for modularity, because [it was deemed](http://mlton.org/References#Wright95) to not only overly complicate the typing system implementation and modular interaction of type signatures in modules, but also could [necessitate sharing typing information between modules](http://archive.is/H8Cyn#selection-2791.566-2791.649) that should be encapsulated.

Annotating whether the type parameter may only be allowed in covariant assignments, and then combining with _value restriction_, _Weak Types_, or _Closure Typing_ in order to obtain fewer restrictions than each of those systems alone. OCaml [has covariant annotations](https://realworldocaml.org/v1/en/html/imperative-programming-1.html#relaxing-the-value-restriction)<sup>2</sup> in the context of a default _value restriction_<sup>3</sup> to overcome restrictions when using abstract types.

Note [_eta expansion_ as a means of ameliorating](http://mlton.org/ValueRestriction#_working_with_the_value_restriction) the restrictions of _value restriction_ causes redundant evaluation of potentially imperative code which can potentially cause unintended side-effects.

Being pure functional, Haskell avoids this typing and restriction problem with closures, but [has the dual of the analogous problems](https://stackoverflow.com/questions/22507448/the-value-restriction#comment34248720_22507665) because it can still model imperativity (otherwise it wouldn’t be a useful PL).

<sup>1</sup> <sub>`f : [A, B, C](A) → (B) → (C) → C`, i.e. a function with closures over each `(A)`, `(B)`, `(C)` of the set of inputs `(A) → (B) → (C)`, as opposed to an uncurried function `f : [A, B, C](A, B, C) → C`</sub>

<sup>2</sup> <sub>Definitionally ‘covariance’ means for example applicable to parametrised type constructors.</sub>

<sup>3</sup> <sub>Thus OCaml [can’t generalize a partially applied](https://stackoverflow.com/questions/22507448/the-value-restriction) (curried) function because it doesn’t have _Closure Typing_ or _Weak Typing_ as a default.</sub>

**Comment by keean** [comment]
For a while I have considered not allowing closures at all, but using function objects. This also emphasises the difference between values and objects (values are immutable and must be copied, objects can be mutable and can be referenced). These things all influence each other, so it is difficult to talk about one in isolation. 

We can lessen the boilerplate by deriving some methods automatically, like a default constructor. Consider a counter, in JS you can write:
```js
function Counter(start) {
   var count = 0
   return {
      inc: () => {...}
      get: () => {...}
   }
}
```
We can write this as a function object, or a module like this:
```
module Counter(start : Int)
   inc()
   get() : Int
private   
   int count
   count := start
```
The main difference is the function object / module does not capture any ambient scoped variables, you have to pass anything you want to use to the constructor. This means the state and size of the object is clearly visible to the programmer. Also when you invoke one it is clear what is being passed in.

Capture of scoped variables is anti-modular.

**Comment by shelby3** [comment]
> Capture of scoped variables is anti-modular.

Disagree. Closures enable more modular composition of functional programming, because for example I can reuse a function which inputs a function that doesn’t have to thread the lexical environment to the input function.

Also for example [a reference constructor is always an implicit closure](https://stackoverflow.com/questions/22507448/the-value-restriction#comment84176376_22507665), so without closures we couldn’t write programs without having to feed in top-down all the memory allocations that would be required. This is analogous to the error in conceptualization you made in the discussion about modularity with @rossberg at LtU in the 1ML thread.

And closures provide encapsulation which another property of modularity.

**Comment by keean** [comment]
> And closures provide encapsulation which another property of modularity, which is important for reasons such as security and enforcing semantic invariants.

See the article on Midori you linked to, ambient authority is considered bad, and is equivalent to capturing outer scope variables. In Midori objects are capabilities, and mutable static object properties are not allowed.

An example of ambient authority would be a "user-id" defined in an outer scope that is referenced by closures defined in that scope.

**Comment by shelby3** [comment]
Now this discussion of modularity is becoming very relevant and interesting to the studying I’ve been doing the past days, but I really wanted to take this up in the thread where we were originally discussing ML modules (will try to move the discussion there eventually).

> Scoped variable capture is anti-modular because it creates a coupling between the code in the function/closure and it's context that prevents that function being used in a different context. This is pretty much the definition of anti-modular.

In some respects that is what modules do when they export an interface then internally encapsulation a non-exported lexical environment. Closures are modules that export a function interface.

One issue is that closures are not applicative at least w.r.t. to memory allocation side-effects. They are also not applicative w.r.t. to the captured lexical environment if any of the environment is mutable (even if not mutated by the closure).

An applicative closure is not prevented from being used in other contexts. Generative modules also have this side-effects dependencies restriction.

> We can only separately compile code that has no coupling to its context.

Agreed but there’s a tradeoff you may not be considering. The lexical environment of the closure has no exported interface, so in this respect closures aren’t modular w.r.t. separate compilation and explicit (contractual) dependencies, but they are modular w.r.t. to encapsulation and being modularly compositional. Whereas, the top-down alternative you advocated was modular w.r.t. separate compilation and explicit (contractual) dependencies, but not as modular w.r.t. to composition and encapsulation.

In essence I presume what @rossberg’s 1ML accomplishes is the unification of module functors with function closures. Yet this requires [predicativity](https://en.wikipedia.org/wiki/Parametric_polymorphism#Predicative_polymorphism) which thus forces a restriction due to the requirement to wrap the impredicative module recursion. Whereas, his MixML unifies signatures and (even partial) implementations (aka structures), thus allowing recursion and the more complex MixML type system apparently allows HKT on implementations instead of only via functors. I’m leaning towards MixML as the better tradeoff but I require more study and @rossberg seems to prefer more compositional 1ML. I expect to eventually figure out that wrapping is as limiting as the non-unification of functors with functions, but I require first to more fully/deeply understand these comparative restrictions and the type systems they require/allow.

> An example of ambient authority would be a "user-id" defined in an outer scope that is referenced by closures defined in that scope.

Agreed that given the lexical environment has no exported signature with encapsulated non-exports, the closure is free to see everything in the lexical environment and leak it outside the function without any signature type checking that leakage.

This btw is afaics another reason that abstract types are modular and type families are not, because type families are injected from outer context and thus limit the degrees-of-freedom on encapsulation joins. The big problem for me (from what I know thus far) with abstract types was [in ML they’re forcing HKT via functors](https://github.com/keean/zenscript/issues/35#issuecomment-360972795) yet MixML seems to remove this restriction with a more complex type system with linear types for tracking the enforced kinds on abstract types (apparently that leaks the kind out of the encapsulation though?). There will also be a tradeoff between the trilogy of encapsulation, composition, explicit signature dependencies. These [“pick 2-of-3 theorems”](https://en.wikipedia.org/wiki/CAP_theorem) seems to arise often in understanding partial orders.

**Comment by keean** [comment]
There are definitely reasons to prefer modules over type classes. Maybe I am starting to prefer the idea of modules + implicits. 

Type families provide a kind of associated types for type-classes. The problem is modular associated types require dependent-types.

Consider a heterogeneous list containing modules with an associated type. If we create a value of the associated type, this type will depend on the module value. Now the p

**Comment by keean** [comment]
Now the problem is we need to emit different code for floats and ints (if we want performance there are different registers and op-codes), but we don't know what type there is until runtime.

I unions can help here, because if we know the union of all modules that this program can pass, then we can monomorphise two versions of the function, and we choose which one based on the module. This would still need to be dynamic/runtime polymorphism.

So this leads back to what I was saying about type propagation, we want to propagate concrete types and unions forward, and type-classes and constraints backwards.

**Comment by shelby3** [comment]
> So this leads back to what I was saying about type propagation, we want to propagate concrete types and unions forward, and type-classes and constraints backwards.

Doesn’t this entirely resolve the selection of specific functors implementing the interfaces at compile-time, except for any higher-rank polymorphism (polymorphic recursion)? The compiler propagates the quantified types down the function call hierarchy and the interface constraints backwards up the hierarchy.

**Comment by sighoya** [comment]
@keean wrote:
> There  are definitely reasons to prefer modules over type classes. Maybe I am starting to prefer the idea of modules + implicits.

Are [modular type classes](https://people.mpi-sws.org/~dreyer/papers/mtc/main-long.pdf) an option?

**Comment by keean** [comment]
@sighoya this is exactly what I was referring to when I said modules+implicits.

The problem is to keep things functional requires applicative modules, and that can lead to problems with associated types. I think you need some kind of dependent types to do this. I would be okay with this if I can find a dependent type system that maintains type erasure.

**Comment by sighoya** [comment]
modular type classes support associated types inside (without dependent types), but I assume, you mean something different. 
What is an applicative module, a functor?

**Comment by keean** [comment]
Modules/Functor can be generative or applicative, an applicative module/functor is like a function and the associated types from an application are always the same. For a generative module/functor each 'new' instantiation generates fresh unique types.

**Comment by keean** [comment]
@sighoya not soundly. A module signature is a type, a module itself is a value. If you pass a module to a function and the type of the function depends on the module, then you have a dependent type.

**Comment by shelby3** [comment]
@keean wrote:
> I would be okay with this if I can find a dependent type system that maintains type erasure.

Can you clarify why monomorphisation can’t be used instead of type erasure? We’re talking typeclass bounds so the instanced type of the abstract type is known at compile-time at the call site.

**Comment by sighoya** [comment]
@keean wrote:
> @sighoya not soundly. A module signature is a type, a module itself is a value. 

Yes, this makes sense. An module can be seen as an instance of a struct as struct instances like modules allow field access. And a struct can be seen as a signature.

 @keean wrote:
>If you pass a module to a function and the type of the function depends on the module, then you have a dependent type.

Okay, but the function signature can also depend on a integer value instead of a module value. So why not disallow it?
Can you present an example, why this is a problem?

**Comment by keean** [comment]
It can be a problem because you cannot always type check the function if you don't know the module value. An integer value is different because it has an integer kind (where a kind is one level up from a type).

For example we have a module signature that has an associated type, say an iterator over a collection where the associated type is the 'value-type' of the collection. We compile some code that uses this module signature we do not know at compile time the type of the 'value-type' because that is not known until runtime when you know the module value that is passed.

**Comment by shelby3** [comment]
> Okay, but the function signature can also depend on a integer value instead of a module value. So why not disallow it?

If I’m understanding the issue correctly, the function takes the signature of a module, then the abstract types in the signature have an unknown instance type (because they’re abstract), because module signatures can be implemented by multiple structures via for example ML functors. Therefore the function is polymorphic.

Additionally it’s

**Comment by keean** [comment]
We can turn the associated type into a type parameter, but this is a problem because a type parameter is an input to a module, and and associated type is an output. By making it an input we lose the encapsulation (the module may represent an array of integers for example, this is a particular issue with iterators that are constructed by collections).

**Comment by shelby3** [comment]
> We can turn the associated type into a type parameter, but this is a problem because a type parameter is an input to a module, and and associated type is an output. By making it an input we lose the encapsulation 

Afaics, even without it being an input we can break encapsulation, because we can request a typeclass bounds (not on the associated type but…) and that typeclass bound and reveal something about the associated typed.

**Comment by sighoya** [comment]
As far as I know, associated types strongly correlate with [existential types](http://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#existentially-quantified-data-constructors) in Haskell. 
Because Haskell does not really (partial) support dependent types, the existential type is not found in the signature of a record, typeclass and so on. It is an internal Variant Type (runtime polymorphism).
Why not simply disallowing associated types in a function/record/module signature?
Maybe a code example can clarify.

**Comment by shelby3** [comment]
> Why not simply disallowing associated types in a function/record/module signature?

They’re required [for implementing multi-parameter typeclasses](https://www.schoolofhaskell.com/school/to-infinity-and-beyond/pick-of-the-week/type-families-and-pokemon).

**Comment by shelby3** [comment]
Please kindly move the modularity discussion to the new Modularity issues thread:

https://github.com/keean/zenscript/issues/39

**Comment by shelby3** [comment]
I’m going to try to incorporate the following in an [ongoing discussion](https://github.com/keean/zenscript/issues/17#issuecomment-366597944) about effects systems.

Erik Meijer [wrote](https://www.infoq.com/interviews/erik-meijer-programming-language-design-effects-purity):

> Q: **But is it a problem if I have some language which has functions, which uses functions but it still does a lot of, for instance IO or whatever other non-pure operations interleaved with functional code?**
> 
> A: I think that’s a huge problem actually and the problem is that once you’ve introduced closures, let’s call them closures because I don’t want to call them functions if I’m really kind of a purist, because they are not functions, they are kind of higher order things, they are like objects, special objects that capture variables in their context. And that’s exactly the problem because when you use a closure the evaluation of that body of the closure gets delayed, it doesn’t happen at that point but it might capture all kind of state or context and it’s not only local variables, it can be other things that it captures, then you pass that closure happily around and at some point you call it, but now when you call it it’s out of the lexical context where it was created and then really bad things can happen. So I think it’s a very powerful tool but you’re giving people an enormous gun and they can shoot their legs off, their heads off, bad things can happen if you don’t know what you’re doing.

**Comment by shelby3** [comment]
A summary of the main issues with closures are:

* [the interplay with](https://github.com/keean/zenscript/issues/32#issuecomment-360765439) the GC and stack activation records
* more error-prone, non-modular spaghetti [to employ them not locally](https://github.com/keean/zenscript/issues/33#issuecomment-362783353) and pass the closure object around distant from the lexical scope where they were instanced. This also relates to [the inadvertent memory allocation](https://github.com/keean/zenscript/issues/33#issuecomment-302353710) which is very costly if escape the generational GC so another reason to employ them only in temporal local contexts.

**Comment by shelby3** [comment]
Referring to my prior post above, I am seeing a [potentially important use case](https://github.com/keean/zenscript/issues/16#issuecomment-415910578) for _non-local_ closures, so I think we will want to allow _implicit_ local closures and non-local closures must be _explicitly_ created.

**Comment by shelby3** [comment]
Referring to my prior post above, noting a [potentially important use case](https://github.com/keean/zenscript/issues/16#issuecomment-415910578) for _non-local_ closures, so perhaps we will want to allow _implicit_ local closures and non-local closures must be _explicitly_ created.

**Comment by sighoya** [comment]
Regarding the scaping problem is this a viable option?:
>https://docs.swift.org/swift-book/LanguageGuide/Closures.html
>https://medium.com/swift-india/functional-swift-closures-67459b812d0
